File: 317-Shortest-Distance-from-All-Buildings/solution.java
public class Solution {
    class Point{
		int x;
		int y;
		public Point(int x, int y){
			this.x=x;
			this.y=y;
		}
	}
	
	boolean isValid(int i,int j, boolean[][]visited,int[][] grid,int m, int n) {		
		if(i<0||i>=m||j<0||j>=n) return false;
		if(grid[i][j] == 2 || grid[i][j] == 1) return false;
		if(visited[i][j]) return false;
		return true;
	}
	
    public int shortestDistance(int[][] grid) {
        HashSet<Point> robot = new HashSet<Point>();
        int m = grid.length, n = grid[0].length;
		//boolean[][] blocker = new boolean[m][n];
		for(int i = 0; i < m; i++) {
		    for(int j = 0; j < n; j++) {
		        if(grid[i][j] == 1) robot.add(new Point(i, j));
		        //if(grid[i][j] == 2) blocker[i][j] = true;
		    }
		}
		int[][] temp = new int[m][n];
		int[][] count = new int[m][n];
        for(Point p:robot) {
			LinkedList<Point> q = new LinkedList<Point>();
			boolean[][] visited=new boolean[m][n];
			q.offer(p);		
			visited[p.x][p.y]=true;
			int len=0;
			while(!q.isEmpty()) {
				int size =q.size();
				for(int i=0;i<size;i++) {
					Point cur = q.poll();					
					temp[cur.x][cur.y]+=len;
					count[cur.x][cur.y]++;
					if(isValid(cur.x+1,cur.y,visited,grid,m,n)) {						
						q.offer(new Point(cur.x+1,cur.y));
						visited[cur.x+1][cur.y]=true;
					}
					if(isValid(cur.x-1,cur.y,visited,grid,m,n)) {
						q.offer(new Point(cur.x-1,cur.y));
						visited[cur.x-1][cur.y]=true;
					}
					if(isValid(cur.x,cur.y-1,visited,grid,m,n)) {
						q.offer(new Point(cur.x,cur.y-1));
						visited[cur.x][cur.y-1]=true;
					}
					if(isValid(cur.x,cur.y+1,visited,grid,m,n)) {
						q.offer(new Point(cur.x,cur.y+1));
						visited[cur.x][cur.y+1]=true;
					}
				}
				len++;
			}
		}
		
		int min=Integer.MAX_VALUE;
		int x = 0, y = 0;
		for(int i=0;i<m;i++) {
			for(int j=0;j<n;j++) {
				if(grid[i][j] == 0 && temp[i][j] < min && count[i][j] == robot.size()) {
				    min=temp[i][j];
				    x = i;
				    y = j;
				}
			}
		}
		return (min == Integer.MAX_VALUE || min == 0) ? -1 : min;
    }
}

File: 317-Shortest-Distance-from-All-Buildings/solution.py
class Solution(object):
    def shortestDistance(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        bs = set()
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    bs.add(i * n + j)
        distance = [[0 for j in range(n)] for i in range(m)]
        count = [[0 for j in range(n)] for i in range(m)]
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        for b in bs:
            queue = [b]
            cur_len = 0
            visited = [[False for j in range(n)] for i in range(m)]
            visited[b / n][b % n] = True
            while queue:
                size = len(queue)
                for s in range(size):
                    cur = queue.pop(0)
                    i, j = cur / n, cur % n
                    distance[i][j] += cur_len
                    count[i][j] += 1
                    for d in directions:
                        x, y = i + d[0], j + d[1]
                        if self.isValid(x, y, m, n, grid, visited):
                            queue.append(x * n + y)
                            visited[x][y] = True
                            
                cur_len += 1
        result = sys.maxint
        for i in range(m):
             for j in range(n):
                 if grid[i][j] == 0 and count[i][j] == len(bs):
                     result = min(result, distance[i][j])
        if result == sys.maxint:
            return -1
        return result
        
    def isValid(self, x, y, m, n, grid, visited):
        if x < 0 or x >= m or y < 0 or y >= n:
            return False
        if visited[x][y]:
            return False
        if grid[x][y] != 0:
            return False
        return True
